### PR TITLE
Abstract out System.Reflection.Emit.ILGenerator

### LINKED ISSUE #86110: [API Proposal] Make System.Reflection.Emit.ILGenerator constructor protected
For new persisting AssemblyBuilder.Save implementation we have abstracted out all other Reflection.Emit public APIs with https://github.com/dotnet/runtime/issues/78542, similar treatment needed for ILGenerator in order to save IL using MetadataBuilder. Literally all public methods of this type are virtual, but the constructor is internal, so I could not implement it outside of CoreLib
```diff
namespace System.Reflection.Emit
{
-   public partial class ILGenerator
+   public abstract class ILGenerator // there is no abstract method, probably make it abstract to avoid creating an instance. 
    {
-       internal ILGenerator() { }
+       protected ILGenerator() { }
        public virtual int ILOffset { get { throw null; } }
        public virtual void BeginCatchBlock(System.Type? exceptionType) { }
        public virtual void BeginExceptFilterBlock() { }
        // ... All public APIs a virtual except one, we can use the `byte arg` overload for this
        [System.CLSCompliantAttribute(false)]
        public void Emit(System.Reflection.Emit.OpCode opcode, sbyte arg) { }
    }
}
```

CC @AaronRobinsonMSFT @jkotas @steveharter

### PR DESCRIPTION
**Pull request by buyaa-n**
Contributes to https://github.com/dotnet/runtime/issues/62956
Fixes https://github.com/dotnet/runtime/issues/86110
- Abstract out ILGenerator, move into shared library
  - Kept some APIs virtual (when implementations in Mono and CLR are same or close, we can add more)
  - Made some public APIs abstract (mostly when the Mono and CLR implementations are different)

### CODE REVIEW

File: src/libraries/System.Reflection.Emit.ILGeneration/ref/System.Reflection.Emit.ILGeneration.cs [hunk 1]
```diff
@@ -13,48 +13,56 @@ public partial class CustomAttributeBuilder
...
-        public virtual void BeginScope() { }
+        protected ILGenerator() { }
+        protected abstract void EnsureCapacity(int size);
+        protected abstract void ILEmit(System.Reflection.Emit.OpCode opcode);
+        protected abstract void WriteByte(byte arg);
+        protected abstract void WriteShort(short arg);
+        protected abstract void WriteInt(int arg);
+        protected abstract void WriteLong(long arg);
+        protected abstract void WriteSingle(float arg);
+        protected abstract void WriteDouble(double arg);
```
Comment (jkotas) [inline src/libraries/System.Reflection.Emit.ILGeneration/ref/System.Reflection.Emit.ILGeneration.cs:26]: These public APIs were not approved. I do not think we should be adding them. Instead, make more of the methods abstract.

Comment (buyaa-n) [reply src/libraries/System.Reflection.Emit.ILGeneration/ref/System.Reflection.Emit.ILGeneration.cs:26]: > The protected constructor and changing the type to abstract suggests that some of the virtual members should themselves flip to abstract (`or that there's a protected abstract missing from the design`). As a general reminder, abstract methods can't be added after the initial release where it becomes possible to actually extend the type, so for ILGenerator that means the current (.NET 8) release.

From the above API review comment I thought `or that there's a protected abstract missing from the design` part imply that we can add `protected abstract` members, anyway, I have no strong opinion to keep them

File: src/mono/System.Private.CoreLib/src/System/Reflection/Emit/DynamicMethod.Mono.cs [hunk 2]
```diff
@@ -59,8 +59,12 @@ public sealed override Delegate CreateDelegate(Type delegateType, object? target
 
         public DynamicILInfo GetDynamicILInfo() => _dynamicILInfo ??= new DynamicILInfo(this);
 
-        public ILGenerator GetILGenerator(int streamSize) =>
-            _ilGenerator ??= new ILGenerator(Module, new DynamicMethodTokenGenerator(this), streamSize);
+        public ILGenerator GetILGenerator(int streamSize) => GetILGeneratorInternal(streamSize);
+
+        internal RuntimeILGenerator GetILGeneratorInternal() => GetILGeneratorInternal(64);
```
Comment (jkotas) [inline src/mono/System.Private.CoreLib/src/System/Reflection/Emit/DynamicMethod.Mono.cs:64]: Nit: `GetRuntimeILGenerator` may be a better name

**Review by jkotas — APPROVED** [review]
LGTM otherwise (once the CI is green).

**Review by jkotas — APPROVED** [review]
LGTM otherwise (once the CI is green).

**Comment by buyaa-n** [comment]
> LGTM otherwise (once the CI is green).

Hm, strangely there are tests that accesses private/internal fields/methods of ILGenerator outside of the CoreLib, not sure how to workaround ... 
https://github.com/dotnet/runtime/blob/78e42f789bd14919918074810069e63d5fd1d8d7/src/libraries/System.Linq.Expressions/tests/ILReader/DynamicMethodILProvider.cs#L13-L19

**Comment by jkotas** [comment]
Change it to access fields of RuntimeILGenerator:
```
```
private static readonly Type s_runtimeILGenerator = Type.GetType("System.Reflection.Emit.RuntimeILGenerator");
private static readonly FieldInfo s_fiLen = s_runtimeILGenerator.GetFieldAssert("m_length");
...
```

**Comment by buyaa-n** [comment]
> Change it to access fields of RuntimeILGenerator:

Thank you 🙏!

**Pull request merged**